### ISSUE TITLE
Fitting integral equation with Gaussian Process

### ISSUE THREAD

**Issue by lujonathanh** [tensorflow/probability#517]
Hello,

Suppose I have data points $X_{1}, X_{2}, X_{3}, X_{3.5}$ (the index indicates the timepoint) from some unknown random process $X_t$. Similarly, points $Y_{1}, Y_{2}, Y_{3}, Y_{3.5}$ from some unknown random process $Y_t$.

I would like to fit the integral equation:

$$Y_t = \int_{0}^{\infty} \beta(\ell) X_{t-\ell} d\ell$$

where $\beta(\ell) \sim GP(0, k_{SE}(., .)$, i.e. $\beta(\ell)$ is a continuous function drawn from a Gaussian Process with squared-exponential kernel.

My goal is to obtain a posterior on $\beta(\ell)$ given my data points $X_{1}, X_{2}, X_{3}, X_{3.5}, Y_{1}, Y_{2}, Y_{3}, Y_{3.5}$.

I was wondering how I might code this model up?

Thank you so much for your help. I really appreciate it!

**Comment by brianwa84** [comment]
I don't know off hand how to (or if you can) integrate a GP. One thought is that you could differentiate the rhs and try using some of the code in tfp.math.ode on the rhs.

**Comment by brianwa84** [comment]
https://math.stackexchange.com/q/1471507 seems loosely relevant, though they only integrate 0 to t.

**Comment by davmre** [comment]
A few thoughts (for posterity at least):

a) As @brianwa84 pointed out, integrals and derivatives of GPs are also GPs under nice conditions. This basically follows from the fact that finite linear combinations of Gaussians are Gaussian, so you can take advantage of the fact that differentiation is basically subtraction (`(f(x+ε) - f(x))/ε`) and integration is basically summation as long as the appropriate limits exist. You can work out the covariances of the integral and derivative GPs as the limits from the finite case, and IIRC you end up getting something in terms of integrals and derivatives (respectively) of the original covariance kernel.

Because all of these quantities are jointly Gaussian, you can even combine observations of a GP-modeled function with observations of its derivative or (finite-interval) integrals. You can use this for fun stuff like Bayesian optimization when you have both function values and gradient information available from autodiff (e.g., https://papers.nips.cc/paper/7111-bayesian-optimization-with-gradients.pdf).

b) I don't *think* the integral of a GP from 0 to \inf will be a nice quantity in general for standard stationary kernels, since you'd expect the variance to go to infinity.

c) Here you're convolving a GP \beta against another unknown function X. That integral will in general depend on values of X at all times, not just times you observe, so it seems like you'll need to make some prior assumptions about the form of X. A natural assumption might be that X is itself a GP. But then the quantity inside the integral is no longer directly a GP, because the product of two  Gaussians RVs is not Gaussian. So it's not clear that you can do this entirely analytically.

d) One way to set this up would be to do approximate inference over a discrete approximation of the convolution kernel represented by X. For any given sampled value for X, the conditional distribution of `y` (and the joint distribution of `(y, beta)`) is then a GP that you can evaluate likelihoods of. This is definitely a not-for-beginners modeling task, but I got a bit nerd-sniped by this and wrote up a (totally untested, use at your own risk) implementation. You'd first need to define a custom covariance kernel for the convolved process `y`, in terms of the underlying `beta` kernel and the (finite set of) points at which you've inferred `X`:

```python
class ConvolvedGPKernel(tfp.math.psd_kernels.PositiveSemidefiniteKernel):
  """Kernel representing a finite sum of points from an underlying GP.
  
  Given an underlying process `f ~ GP(0, k)` with kernel `k`, this is the
  kernel of the process convolved (finitely) with `weights`, i.e.,
  `g[t] = [sum(f[t+offsets[i]] * weights[i] for i in range(len(offsets))] / normalizer`
  """
  def __init__(self, base_kernel, offsets, weights, normalizer=1.):
    self._base_kernel = base_kernel
    self._offsets = offsets
    self._weights = weights
    self._weights_outer_product = weights[tf.newaxis, :] * weights[:, tf.newaxis]
    self._num_points = tf.shape(offsets)[-1]  # Assume == shape(weights)[-1].
    self._normalizer = normalizer

  def _apply(self, x1, x2, example_ndims=0):
  #   Computes the sum of covariances, cov(g[x1], g[x2]) = sum([
  #       cov(f[x1+offsets[i1]] * weights[i1], f[x2+offsets[i2]] * weights[i2])
  #       for i1 in range(num_points) for i2 in range(num_points)]) /  normalizer**2
  expanded_x1 = x1[tf.newaxis, tf.newaxis, ...] * tf.reshape(
    self._offsets, tf.concat([[-1, 1], tf.ones(tf.shape(x1))], axis=0)
  expanded_x2 = x2[tf.newaxis, tf.newaxis, ...] * tf.reshape(
    self._offsets, tf.concat([[1, -1], tf.ones(tf.shape(x2))], axis=0)
  all_kernels = self._base_kernel(expanded_x1, expanded_x2, example_ndims=example_ndims)
  weighted_kernels = all_kernels * tf.reshape(
      self._weights_outer_product,
      tf.concat([tf.shape(self._weights_outer_product), tf.ones(tf.shape(all_kernels)[2:])], axis=0))
  return tf.reduce_sum(weighted_kernels, axis=(0, 1)) / self._normalizer**2
```

Then you could write a model like this:

```python
Root = tfd.JointDistributionCoroutine.Root
def model_fn():
  num_points = 100
  tmin, tmax = -5., 5.
  index_points = tf.linspace(tmin, tmax, num)_points)  # Adjust as appropriate.
  X = yield Root(tfd.GaussianProcessRegressionModel(kernel=some_kernel_fn,
                                                    observation_noise_variance=0.1,  # Or learn this.
                                                    observation_index_points=X_timepoints, 
                                                    observations=X_observed_values,
                                                    index_points=index_points))
  #  X is a vector of length `num_points` representing the inferred convolution kernel.
  # We approximate `y[t] = sum([beta[t+i] * X[i] for i in index_points]) * (tmax-tmin)/num_points`.
  y_kernel = ConvolvedGPKernel(base_kernel=beta_kernel,
                               offsets=index_points,
                               weights=X,
                               normalizer=(tmax-tmin)/num_points)
  Y = yield tfd.GaussianProcess(kernel=y_kernel,
                                observation_noise_variance=0.1,  # Or learn this.
                                index_points=Y_timepoints)
model = tfd.JointDistributionCoroutine(model_fn)
```

At this point you'd condition on the observed Y's (`target_log_prob = lambda x: model.log_prob((x, Y_observed_values)`) and fit the latent X's using the approximate inference method of your choice (VI, HMC, HUTS). 

To get out the values for the underlying betas, you'd then need to extend the ConvolvedGPKernel to represent a joint kernel over the original and convolved values (`f` and `g`), and pass this into a `GaussianProcessRegressionModel` that observes the `Y` values and predicts values at `beta` index points. This is a bit tricky since you'd need a way to distinguish points in `f`-space from points in the convolved `g`-space. One way to hack this would be to use an extra indicator dimension, so `x = [0, 2.1]` would index to the value `f(2.1)` and `x = [1, 2.1]` to `g(2.1)`; and then you'd use something like `tf.where` in the kernel to choose the appropriate covariance for each pair of inputs. That'd be messy, but I think it should work (let us know if you'd like help going this route).

As a side note, something like `ConvolvedGPKernel` would be a great contribution to TFP proper. :-)

**Comment by lujonathanh** [comment]
Wow. Thank you both so much for following up and writing this so thoroughly-- I really appreciate it. I'm going to try this out, though unfortunately only until about a month from now. I will let you know if I have any followups.

This is such a wonderful and supportive community, and I truly appreciate your time. Thank you!!

**Comment by srvasude** [comment]
Closing this issue. Please reopen if you have any further questions.

There is also now a tfp.ode package as well. It's possible some parts of it might be useful in your work if you are doing more exotic GP-like things (We don't have SDE support, but it might be possible to reuse one of the integrators in the package for integrating your SDE).
